plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.hs

```diff
@@ -46,213 +65,350 @@ data TypeBuiltinG
-extendSub s (FS i) = renameTypeG FS (s i)
-
--- |Simultaneous substitution of variables in generated types.
-substTypeG :: (n -> TypeG m) -> TypeG n -> TypeG m
-substTypeG s (TyVarG i)             = s i
-substTypeG s (TyFunG ty1 ty2)       = TyFunG (substTypeG s ty1) (substTypeG s ty2)
-substTypeG s (TyIFixG ty1 k ty2)    = TyIFixG (substTypeG s ty1) k (substTypeG s ty2)
-substTypeG s (TyForallG k ty)       = TyForallG k (substTypeG (extendSub s) ty)
-substTypeG _ (TyBuiltinG tyBuiltin) = TyBuiltinG tyBuiltin
-substTypeG s (TyLamG ty)            = TyLamG (substTypeG (extendSub s) ty)
-substTypeG s (TyAppG ty1 ty2 k)     = TyAppG (substTypeG s ty1) (substTypeG s ty2) k
+-- ** Type checking
+
+checkTypeG
+  :: Eq tyname
+  => KCS tyname
+  -> TCS tyname name
+  -> TypeG tyname
+  -> TermG tyname name
+  -> Cool
+checkTypeG _ tcs ty (VarG i)
+  = varTypeOk
+  where
+    varTypeOk = toCool $ ty == typeOf tcs i
+
+checkTypeG kcs tcs (TyForallG k ty) (TyAbsG tm)
+  = tmTypeOk
+  where
+    tmTypeOk = checkTypeG (extKCS k kcs) (firstTCS FS tcs) ty tm
+
+-- NOTE: In the PlutusCore type checker, the type of the body is
+--       not necessarily in normal form. I've opted to force all
+--       types to be in normal form, so there's no normalization.
+--
+checkTypeG kcs tcs (TyFunG ty1 ty2) (LamAbsG tm)
+  = tyKindOk &&& tmTypeOk
+  where
+    tyKindOk = checkKindG kcs (Type ()) ty1
+    tmTypeOk = checkTypeG kcs (extTCS ty1 tcs) ty2 tm
+
+checkTypeG kcs tcs ty2 (ApplyG tm1 tm2 (Normalized ty1))
+  = tm1TypeOk &&& tm2TypeOk
+  where
+    tm1TypeOk = checkTypeG kcs tcs (TyFunG ty1 ty2) tm1
+    tm2TypeOk = checkTypeG kcs tcs ty1 tm2
+
+checkTypeG kcs tcs vTy (TyInstG tm (Normalized vCod) (Normalized ty) k)
+  = tmTypeOk &&& tyKindOk &&& tyOk
+  where
+    tmTypeOk = checkTypeG kcs tcs (TyForallG k vCod) tm
+    tyKindOk = checkKindG kcs k ty
+    tyOk = vTy == normalizeTypeG (TyAppG (TyLamG vCod) ty k)
+
+checkTypeG _ _ _ _ = false
+
+
+instance Check ClosedTypeG ClosedTermG where
+  check = checkTypeG emptyKCS emptyTCS
+
+instance Check (Kind (), ClosedTypeG) ClosedTermG where
+  check (_k, tyG) = check tyG
+
+
+-- ** Type checking state
+
+newtype TCS tyname name = TCS{ typeOf :: name -> TypeG tyname }
+
+emptyTCS :: TCS tyname Z
+emptyTCS = TCS{ typeOf = fromZ }
+
+extTCS :: forall tyname name. TypeG tyname -> TCS tyname name -> TCS tyname (S name)
+extTCS ty TCS{..} = TCS{ typeOf = typeOf' }
+  where
+    typeOf' :: S name -> TypeG tyname
+    typeOf' FZ     = ty
+    typeOf' (FS i) = typeOf i
+
+firstTCS :: (tyname -> tyname') -> TCS tyname name -> TCS tyname' name
+firstTCS f tcs = TCS{ typeOf = fmap f . typeOf tcs }
+
+
+-- * Normalisation
+
+-- ** Type reduction
+
+type TySub n m = n -> TypeG m
+
+-- |Extend type substitutions.
+extTySub :: TySub n m -> TySub (S n) (S m)
+extTySub _ FZ     = TyVarG FZ
+extTySub s (FS i) = FS <$> s i
+
+
+-- |Simultaneous substitution of type variables.
+applyTySub :: (n -> TypeG m) -> TypeG n -> TypeG m
```

Comment:
You could get substitution and renaming for free by using [`bound`](https://hackage.haskell.org/package/bound-2.0.1/docs/Bound.html) (see [this](https://github.com/input-output-hk/plutus/pull/2077#issuecomment-668002993) comment for how to integrate it with NEAT).

Reply:
I don't agree that you get them for free exactly. When using bound you have to define `>>=` as part the monad instance which *is* substitution. One can do the same for the present approach:
```
instance Monad TypeG where
  a >>= f = applyTySub f a
  return = pure

instance Applicative TypeG where
  (<*>) = ap
  pure = TyVarG
```

Renaming does come for free either way. I think what you do gain is that avoiding some boilerplate for `Scope`. But, `Scope` is harder to understand for bound. I am still on the fence I think.

Reply:
Yet another option is to track contexts like `metatheory` with a third constructor `T` so we only need one context i n term case.
 
Something like this:

```
data Z
  deriving (Typeable, Eq, Show)

data S n
  = SZ
  | SS n
  deriving (Typeable, Eq, Show, Functor)

data T n
  = TZ
  | TS n
  deriving (Typeable, Eq, Show, Functor)

data TermG name
    = VarG
      name
    | LamAbsG
      (TermG (S name))
    |  TyAbsG
      (TermG (T name))
```

This would be my preferred option if it worked but it would make it harder to convert to plutus-core terms so there's no point at the moment I don't think.

Reply:
> I don't agree that you get them for free exactly. When using bound you have to define `>>=` as part the monad instance which is substitution

Normally, you can [derive it](https://hackage.haskell.org/package/bound-2.0.1/docs/Bound.html#v:makeBound).

Reply:
Ah, I see, got the wrong impression from higher up on that page.